### DISCUSSION TITLE
Derive interface default implementations

### DISCUSSION THREAD

**Discussion by msedi** [dotnet/csharplang#4380] [Q&A] [answered]
I'm having following situation where I'm not sure if it's possible:

```cs
using System;

namespace ConsoleApp3
{
    class Program
    {
        static void Main(string[] args)
        {
            Console.WriteLine("Hello World!");
        }
    }

    public interface IArray3D
    {
        int SizeX { get; }
        int SizeY { get; }
        int SizeZ { get; }
        public abstract int[] GetData();
    }

    public interface IVolume : IArray3D
    {
        int[] GetSlice(int index);

        public new int[] GetData()
        {
            int[] result = new int [SizeX*SizeY*SizeZ];
            for (int i = 0; i < SizeZ; i++)
            {
                GetSlice(i).AsSpan().CopyTo(result.AsSpan().Slice(i * SizeX * SizeY, SizeX * SizeY));
            }

            return result;
        }
    }

    public class Volume1 : IVolume
    {
        public int SizeX => throw new NotImplementedException();

        public int SizeY => throw new NotImplementedException();

        public int SizeZ => throw new NotImplementedException();

        public int[] GetData()
        {
            throw new NotImplementedException();
        }

        public int[] GetSlice(int index)
        {
            throw new NotImplementedException();
        }
    }
}
```

The explanation is pretty simple: I have a lot of classes the implement the interface `IArray3D`. Each `IArray3D` has a GetData(). 
Additionally, I have a `IVolume` "derived from `IArray3D`. The `IVolume` has a default implementation of how to combine individual slices into a "total volume array".

So far so good, the `IVolume` is implemented by many different `Volume`s but if I do it in the way described above the compiler complains to implement also the `GetData()` of `IArray3D`.

So my question is if this is possible at all? Of course the option would be to create a helper method for each individual `Volume` class and implement it in each Volume class separately, event more, I would replicate code many times. So I though it would make more sense to do it as descibed above if it would work.

**Comment by yaakov-h** [comment, marked as answer]
I think you can do it with explicit interface implementation, but your code contains numerous compilation errors that its hard to try and reason with it into something that actually achieves your aims.

Can you please remove all compiler errors from your sample code except for the one indicating your problem?

**Reply by msedi** [reply]
@yaakov-h : Thanks for the criticism ;-) You are right, I just wrote down the code just to visualize my problems without checking for errors. I have updated my example so that it works. Thanks for your help.

**Reply by svick** [reply]
> I think you can do it with explicit interface implementation

You're right. After changing the declaration of `GetData` in `IVolume` to the following, the code still compiles even after I remove `GetData` from `Volume1`:

```c#
int[] IArray3D.GetData()
```

**Reply by msedi** [reply]
@svick: Thanks. I'm not quite sure if I understood your change. Did you modify something in IVolume or Volume? I tried a few constellations but it doesn't work. It still asks for an interface implementation.

**Reply by svick** [reply]
@msedi 

Here is the modified code that compiles for me:

```c#
public interface IVolume : IArray3D
{
    int[] GetSlice(int index);

    int[] IArray3D.GetData()
    {
        int[] result = new int[SizeX * SizeY * SizeZ];
        for (int i = 0; i < SizeZ; i++)
        {
            GetSlice(i).AsSpan().CopyTo(result.AsSpan().Slice(i * SizeX * SizeY, SizeX * SizeY));
        }

        return result;
    }
}

public class Volume1 : IVolume
{
    public int SizeX => throw new NotImplementedException();

    public int SizeY => throw new NotImplementedException();

    public int SizeZ => throw new NotImplementedException();

    public int[] GetSlice(int index)
    {
        throw new NotImplementedException();
    }
}
```

**Reply by msedi** [reply]
@svick : Thanks for your help. Something is really wrong with my environment. Even when changing the things it didn't compile. After shutting down and opening again it works.
